### ISSUE TITLE
Guard against MitM attack: remove node name from the token

### ISSUE THREAD

**Issue by louiseschmidtgen** [canonical/microcluster#96]
I propose enhancing the security of microcluster join tokens by removing the node name from the token payload. Currently, including the node name in the token compromises the validation process, particularly in guarding against potential attacks.

Utilizing Angelos's previous work as a starting point ([neoaggelos:microcluster:dev/token-names](https://github.com/canonical/microcluster/compare/main...neoaggelos:microcluster:dev/token-names?expand=1)), I suggest revising the client auth.

**Comment by louiseschmidtgen** [comment]
@neoaggelos @tomponline @bschimke95

**Comment by tomponline** [comment]
@louiseschmidtgen please could you describe a scenario where the node name in the token payload could be abused? What impact would that have? Thanks

@masnax i'd be interested to hear your thoughts on this? Is cluster member name only taken from the token?

**Comment by bschimke95** [comment]
Adding some more context here: 
Currently, microcluster does not validate that the node is actual the one that we issued a token for. If an malicious actor obtains the join token it can join the cluster without further validation. For example, this is possible in the k8s snap:
```
# Token is generated on the cluster node
clusternode$ sudo k8s get-join-token mygoodnode
<token>

# Bad actor gets <token>, decodes it, gets node name and can now join some random other node:
mybadnode$ sudo k8s join-cluster <token> --name mygoodnode 
```
This works. The node name that is part of the token payload is not used at all in microcluster right now.

So, if a bad actor gets the token he can just read the payload and get the node name to join the cluster. 
To avoid that, we should remove the node name from the token payload and use the node name that the joining node provides to validate that the node uses the token that was issued for it.

**Comment by tomponline** [comment]
I see thanks. So the main complaint is that it is possible to derive which cluster member the join token was issued for from the join token itself and thus makes it easier to use said join token if a malicious actor gets hold of it.

The join tokens themselves are meant to be secret - they represent a way to join a cluster so should not be made public.
However if the cluster member name inside the token isn't used then it would see sensible to remove it and make things more difficult for an attacker to use a join token if they get hold of one.

**Comment by beliaev-maksim** [comment]
@bschimke95 will node name be sufficient ? This could be easily spoofed 

anyway implement real fingerprint validation where user can compare fingerprints on both main and worker nodes

**Comment by beliaev-maksim** [comment]
@tomponline 
I could be mistaken, but it seems like the current approach protects against compromised or mocked clusters from accepting nodes. However, it doesn't guard against the opposite scenario where an incorrect node attempts to join a valid cluster.

Is join confirmation already planned for implementation in the roadmap?

**Comment by tomponline** [comment]
@beliaev-maksim 

> I could be mistaken, but it seems like the current approach protects against compromised or mocked clusters from accepting nodes. However, it doesn't guard against the opposite scenario where an incorrect node attempts to join a valid cluster.

I would like to have @masnax confirmation on that as he has been most involved with this project in the early days.

> Is join confirmation already planned for implementation in the roadmap?

Yes thats right, @roosterfish is assigned to start a spec for this next.

**Comment by setharnold** [comment]
Would sticking `SECRET` or `PRIVATE` in the token in place of the node name be enough to communicate to administrators that these tokens must be protected?

**Comment by tomponline** [comment]
The token itself is encoded (base64 iirc) so unless you're actively decoding it wouldn't make any difference to an end user admin. The docs might need improving to make it clear thus is a secret token (effectively a one use password).

**Comment by tomponline** [comment]
Join tokens were added to LXD  as a more secure alternative to the long lived trust password which was historically required to join a cluster. The join tokens replaced this with a one time random password and automated verifying the cluster  being joined was the right one when establishing trust.

**Comment by bschimke95** [comment]
@tomponline Thanks for clearing that up. It'd be awesome to be part of this process. Having those security features in microcluster would save us some hassle instead of building them ourselves.

@roosterfish, please include the  Kubernetes team in the spec review, thanks!

About the timeline, you mentioned it should be done before MicroCloud LTS. When's that scheduled for?

**Comment by tomponline** [comment]
@bschimke95 its dependent on getting the MicroCloud token join stuff completed and on when MicroCeph Reef gets out.

Plan is for around last days of May, first days of June currently.

**Comment by masnax** [comment]
Sorry for being late getting to this, I was away last week. the `Name` field in the token is a remnant from early in microcluster's development and it's entirely unused now. It's assigned in only 2 places (and one of those places was actually recently introduced by @neoaggelos)

I see no issue in removing the field, though such a change would be blocked on PR #86 to ensure proper detection across mismatched microclusters.

**Comment by tomponline** [comment]
Hi all,

Our initial specification describing approaches to improve the microcloud cluster join process is available for review:

https://discourse.ubuntu.com/t/secure-cluster-join-in-microcloud/44261

Thanks

**Comment by niemeyer** [comment]
@tomponline Thanks Tom. I'm unable to login on the forum right now due to an issue with SAML there, so will reply here meanwhile.  I would appreciate having a conversation on the topic before this is final, as I'm not sure I correctly understand the requirements and the issues we're attempting to solve.  For example, it strikes me as odd that in an attempt to fix a supposed MitM attack scenario, we're actually introducing interactivity which wasn't there before, and that interactively actually seems to present data from untrusted third-parties that was not possible on a simpler secure-token based approach.  It also seems to disregard the reality of people that are doing this on a regular basis with automation.  I'm hoping we can address whatever shortcomings we may find at the moment without actually making matters worse.

**Comment by tomponline** [comment]
Thanks @niemeyer. Yes I would appreciate your thoughts on this too, we're not going to make a final decision before Madrid as we want to ensure the other `micro*` teams are happy with the changes going into `microcluster`, and in general the join process that MicroCloud will offer, so that hopefully standalone installs of `micro*` can offer similar join processes when not being orchestrated by MicroCloud.

> I'm not sure I correctly understand the requirements and the issues we're attempting to solve. For example, it strikes me as odd that in an attempt to fix a supposed MitM attack scenario, we're actually introducing interactivity which wasn't there before, and that interactively actually seems to present data from untrusted third-parties that was not possible on a simpler secure-token based approach.

There are 2 tangentially related issues going on. The first is the MitM attack that this GH issue is seeing to address, and the other is a request from Mark Shuttleworth to make the MicroCloud join process interactive such that it requires positive confirmation on both the initialising node and the joining nodes with a mindset of "Don't trust the LAN".

This originally came up because he expressed concern that it was possible to deploy MicroCloud in an automated fashion without confirmation on the joining nodes.

I share your concerns about using a long lived, persisted, shared password for signing the MDNS broadcasts (option C in the spec) as that starts to look similar to LXD's trust password option (which was superseded by its join token concept). 

I've actually just finished my 1:1 with @roosterfish and in it we discussed 2 variations on option B and C both seeking to avoid using a long-lived, persisted shared password.

In option B2, we're suggesting using a per-session join token (similar to the per-node join tokens we have today, but with node name removed) which could be used to join multiple nodes within a single session. The join token would be generated on `microcloud init` and then manually (or via external orchestrator) copied to the joining nodes which would pass it to `microcloud join`. When any of the `microcloud join` or `microcloud init` processes end they would forget the session level join token. This would allow a single join token to be used with multiple nodes, but only for the duration that `microcloud init` is being run. This would also remove the need for MDNS entirely as the join token would continue to have the initiators address and cluster certificate in it for verification and to send the join request.

One drawback of copying join tokens between servers has been flagged by @simondeziel is that it can be awkward to do this if you are switching between servers on a datacenter KVM system that doesn't have copy/paste between screens.

So option C2 would seek to address this by using a shorter per-session random join password (rather than a long lived one), which the operator would copy/type into each `microcloud join` process. This would be used to send an HMAC signed MDNS broadcast from the joining nodes, which would be picked up by the `microcloud init` process. In this way the password isn't being broadcast in the clear across the network (as it is today in the MDNS broadcast from joining nodes).

Option C2 also proposes to avoid sending out the MDNS broadcast all of the time before a node is joined and instead would only start to broadcast once `microcloud join` has been started with the shared-password. In this way both B2 and C2 would require the "pairing" process to be started on each joining node first. Avoiding the remote-take over behaviour we have today.

Both interactive join options would also display the joining server's certificate fingerprint on the joining node and the initiator node for manual comparison to allow checking that it is the node that is expected.

My personal preference at this time is option B2 as it removes the need for MDNS (HMAC signed or not) entirely. 
But I acknowledge this would potentially be an issue if setting up nodes on a KVM console.

It could also be extended to per-node join tokens by either allowing an operator to enter the joining node names manually and have `microcloud init` issue join tokens for each node, or by using MDNS to collect the unconfigured nodes and allow the operator to pick which ones to issue join tokens for.

This would require more work from the operator, but would shorten the life of each join token further to single use.

>  It also seems to disregard the reality of people that are doing this on a regular basis with automation. I'm hoping we can address whatever shortcomings we may find at the moment without actually making matters worse.

I believe the original intent of the fully-automated join process was to make it easy to deliver pre-installed devices in to factories/shops and allow simple setup without needing to configure remote console or SSH access to every member in the cluster.

But this is one aspect of the proposed changes we need to get further clarity on the requirements because as you point out the requirement for interactivity seems to be at odds with original design intent and could reduce convenience when deploying into remote environments.
